### PR TITLE
integrating lsl

### PR DESCRIPTION
**Pull request by Yury-Shevchenko**
jsPsych experiment pushes a value into the created lsl stream on each on_start event. 
To subscribe to the stream, I have used the following code in a separate application running on a different computer in the same network:

const lsl = require('node-lsl');
const streams = lsl.resolve_byprop('type', 'Marker');
streamInlet = new lsl.StreamInlet(streams[0]);
streamInlet.streamChunks(12, 1000);
streamInlet.on('chunk', console.log);
streamInlet.on('closed', () => console.log('LSL inlet closed'));

### CODE REVIEW

File: app/utils/jspsych/functions.js [hunk 1]
```diff
@@ -12,6 +13,22 @@ import {
   ExperimentParameters
 } from '../../constants/interfaces';
 
+// create and open an outlet to stream data
+const info = lsl.create_streaminfo(
+  'Dummy',
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: `Dummy` --> `jsPsych`

File: app/utils/jspsych/functions.js [hunk 2]
```diff
@@ -12,6 +13,22 @@ import {
...
+  'Marker', // type
+  1, // number of channels
+  10, // sample rate in 1 sec.
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: based on this [example](https://github.com/labstreaminglayer/liblsl-Python/blob/master/pylsl/examples/SendStringMarkers.py), I would suggest `10` -> `0`.

File: app/utils/jspsych/functions.js [hunk 3]
```diff
@@ -12,6 +13,22 @@ import {
...
+  lsl.channel_format_t.cft_float32,
+  'Dummy Marker Device'
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: `Dummy Marker Device` -> `string`

File: app/utils/jspsych/functions.js [hunk 4]
```diff
@@ -12,6 +13,22 @@ import {
...
+);
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: I would add the final argument `jspsych_brainwaves` for the string identifier. this way we can resolve the stream using this `source_id`.

File: app/utils/jspsych/functions.js [hunk 5]
```diff
@@ -12,6 +13,22 @@ import {
...
+const desc = lsl.get_desc(info);
+lsl.append_child_value(desc, 'manufacturer', 'Random Inc.');
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: for manufacturer, you can say `BrainWaves`

File: app/utils/jspsych/functions.js [hunk 6]
```diff
@@ -12,6 +13,22 @@ import {
...
+const channels = lsl.append_child(desc, 'channels');
+const channel = lsl.append_child(channels, 'channel');
+lsl.append_child_value(channel, 'type', 'Marker');
+const outlet = lsl.create_outlet(info, 0, 360);
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: just curious, what does the `360` refer to?

File: app/utils/jspsych/functions.js [hunk 7]
```diff
@@ -123,11 +140,20 @@ export const instantiateTimeline = (
         if (trial.id === 'trial') {
           return {
             ...trial,
-            on_start: () =>
+            on_start: e => {
+              if (e) {
+                const correctResponse = parseFloat(e.correct_response);
```
Comment (teonbrooks) [inline app/utils/jspsych/functions.js]: for the on_start, is `e` the callback event? I think that we would like to have the eventType passed as the marker to the LSL string similar to what the `eventCallback` is doing below.

File: app/utils/jspsych/functions.js (hunk 2 above)
Comment (Yury-Shevchenko) [reply app/utils/jspsych/functions.js]: Thank you for pointing out to this example.

File: app/utils/jspsych/functions.js (hunk 6 above)
Comment (Yury-Shevchenko) [reply app/utils/jspsych/functions.js]: In node-lsl documentation, it's "max_buffered"
https://github.com/urish/node-lsl/blob/master/index.d.ts#L50

File: app/utils/jspsych/functions.js (hunk 7 above)
Comment (Yury-Shevchenko) [reply app/utils/jspsych/functions.js]: Thanks, I see the idea now. In the new version, eventType is passed as a float number. I have not found support for string in node-lsl. Shall I work on adding it to the node-lsl library, or leave like this with float numbers?

File: app/utils/jspsych/functions.js (hunk 3 above)
Comment (Yury-Shevchenko) [reply app/utils/jspsych/functions.js]: I have not found support for strings streaming in node-lsl (see the last comment).

File: app/utils/jspsych/functions.js (hunk 7 above)
Comment (teonbrooks) [reply app/utils/jspsych/functions.js]: I think leaving it to just float numbers should be fine

File: app/utils/jspsych/functions.js (hunk 6 above)
Comment (teonbrooks) [reply app/utils/jspsych/functions.js]: got it, thanks

**Comment by teonbrooks** [comment]
cool, this looks good. 

after chatting with @jdpigeon, we discussed that we would like to this to be a pref based on the experiment creation. with the custom experiments, we have a checkbox to determine whether it will use eeg or not. we would like to add a checkbox for whether we will use LSL or not. then we will use that state to pref it on here.

this isn't a high priority right now, just wanted to give some feedback on what we were thinking

**Comment by Yury-Shevchenko** [comment]
yes, i think it totally makes sense to connect the LSL functionality with UI and let the user decide.

**Comment by sdikker** [comment]
yes this would be great! we're looking into integrating other physio measures besides EEG that may be easier to collect (e.g., HR) that are LSL compatible

**Pull request closed without merging**